CHANGELOG.md

```diff
@@ -2,13 +2,23 @@
 
 ### Breaking changes
 
-* `RealmResults.distinct()` returns a new `RealmResults` object instead of filtering on the original object.
+* `RealmResults.distinct()` returns a new `RealmResults` object instead of filtering on the original object (#2947).
+
+### Enhancements
+
+* Add support for sorting by link's field (#672).
 
 ### Object Server API Changes (In Beta)
 
 * Add a default `UserStore` based on the Realm Object Store (`ObjectStoreUserStore`).
 * Change the order of arguments to SyncCredentials.custom to match iOS: token, provider, userInfo
 
+### Internal
+
+* Use Object Store's `Results` as the backend for `RealmResults` (#3372).
+  - Use Object Store's notification mechanism to trigger listeners.
+  - Local commit triggers Realm global listener and `RealmObject` listener on current thread immediately instead of in the next event loop.
```

Comment:
@Zhuinden the OS's `SharedRealm::refresh()` is doing the same as a Realm get notified by remote changes. This means we can enable listeners on non-looper thread and they will be triggered when calling `waitForChange()` on non-looper thread Realm. I did enabled that in this PR, but found there are a corner case cannot be fixed easily: 1. we only expose `waitForChange()` 2. we are calling `refresh()` in `waitForChange()` 3. if the the db version has not been advanced, the `refresh()` is not called. 4. so --- how do we deliver async query notification if the db version does not change?

I am still not quite convinced by the `refresh()` is useful (especially it is quite useless in looper thread.). But sometimes i found `refresh()` is quite a good excuse to answer support questions: your realm doesn't change? just call `refresh()`! more times! Although the right thing to do is relying on the change listeners ... 

Calling those callback immediately is probably the better way to deal with local transactions :P